packages/compass-shell/src/modules/worker-runtime/runtime.js

```diff
@@ -0,0 +1,130 @@
+import { createCaller, exposeAll } from './rpc';
+
+function createChildFromSource(src) {
+  const tmpFile = path.resolve(
+    (electron.app || electron.remote.app).getPath('temp'),
+    `worker-runtime-${crypto.randomBytes(32).toString('hex')}`
+  );
+
+  // eslint-disable-next-line no-sync
+  fs.writeFileSync(tmpFile, src, 'utf8');
+
+  const childProcess = child_process.spawn(process.execPath, [tmpFile], {
+    stdio: ['inherit', 'inherit', 'inherit', 'ipc'],
+    env: { ...process.env, ELECTRON_RUN_AS_NODE: 1 },
+    /**
+     * This would be awesome, but only supported since node 12.16.0, compass
+     * is currently at 12.4
+     * @see https://nodejs.org/api/child_process.html#child_process_advanced_serialization
+     */
+    serialization: 'advanced',
+  });
+
+  return childProcess;
+}
+
+class WorkerRuntime {
+  constructor(uri, options, cliOptions) {
+    this.initOptions = { uri, options, cliOptions };
+    this.evaluationListener = null;
+    this.cancelEvaluate = () => {};
+    this.initWorker();
+  }
+
+  initWorker() {
+    const { uri, options, cliOptions } = this.initOptions;
+    this.childProcess = createChildFromSource(worker);
+    this.worker = createCaller(
+      ['init', 'evaluate', 'getCompletions'],
+      this.childProcess
+    );
+    this.workerEvaluationListener = exposeAll(
+      {
+        onPrint: (...args) => {
+          if (this.evaluationListener && this.evaluationListener.onPrint) {
+            this.evaluationListener.onPrint(...args);
+          }
+        },
+        onPrompt: (...args) => {
+          if (this.evaluationListener && this.evaluationListener.onPrompt) {
+            this.evaluationListener.onPrompt(...args);
+          }
+        },
+        toggleTelemetry: (...args) => {
+          if (
+            this.evaluationListener &&
+            this.evaluationListener.toggleTelemetry
+          ) {
+            this.evaluationListener.toggleTelemetry(...args);
+          }
+        },
+      },
+      this.childProcess
+    );
+    this.initPromise = this.worker.init(uri, options, cliOptions);
+  }
+
+  async evaluate(code) {
+    await this.initPromise;
+
+    const result = await Promise.race([
+      this.waitForCancelEvaluate(),
+      this.worker.evaluate(code),
+    ]);
+
+    if (result && result.__error) {
+      const error = new Error();
+      Object.assign(error, result.__error);
+      throw error;
+    }
+
+    return result;
+  }
+
+  async getCompletions(code) {
+    await this.initPromise;
+
+    const result = await this.worker.getCompletions(code);
+
+    if (result && result.__error) {
+      const error = new Error();
+      Object.assign(error, result.__error);
+      throw error;
+    }
+
+    return result;
+  }
+
+  setEvaluationListener(listener) {
+    const prev = this.evaluationListener;
+    this.evaluationListener = listener;
+    return prev;
+  }
+
+  terminateWorker() {
+    this.cancelEvaluate();
+    this.childProcess.kill();
```

Comment:
So currently we are terminating the whole process and starting it from scratch, but it can be definitely improved in the future. `vm` that is running our code can be interrupted by `SIGINT`, it just doesn't work correctly in the Electron version that we are running. There are more details about that in technical design doc

Reply:
Hm yeah … do we want to do the Worker-inside-child-process thing to work around this? Do you want me to ping the electron team about what they can do in terms of adding the Node.js patch to their releases (I’m afraid that would still mean us having to bump the electron version, but maybe that’s okay because it’s not like we can put that off forever…)?

Reply:
Just added an implementation for that in 37538dc, works great!